doc/design/determinism_annotation.md

```diff
@@ -0,0 +1,58 @@
+# Determinism Annotations
```

Comment:
@MatthiasKillat It might be a good idea to move this into `doc/website/concepts/` and also link to this document from `CONTRIBUTING.md`.

Reply:
@mossmaurice I agree, but this is experimental and was hastily scribbled together to see whether we want to go this way. It is in no way mature (does not need to be until some decisions are made).